configuration_broker/config_broker.cc

```diff
@@ -0,0 +1,220 @@
+#include <vector>
+
+#include "config_broker.h"
+#include "sandbox.h"
+
+// Import some useful things from the CHERI namespace.
+using namespace CHERI;
+
+// Expose debugging features unconditionally for this compartment.
+using Debug = ConditionalDebug<DEBUG_CONFIG_BROKER, "Config Broker">;
+
+// Internal represention of Configurtaion Items
+struct NamedConfigItem
+{
+	char      *name; // name matched to capabilities
+	ConfigItem item;
+};
+
+//
+// Set of config data items.
+//
+std::vector<struct NamedConfigItem *> configData;
+
+//
+// unseal a config capability.
+//
+ConfigToken *config_capability_unseal(SObj sealedCap)
+{
+	auto key = STATIC_SEALING_TYPE(ConfigKey);
+
+	ConfigToken *token =
+	  token_unseal<ConfigToken>(key, Sealed<ConfigToken>{sealedCap});
+
+	if (token == nullptr)
+	{
+		Debug::log("invalid config capability {}", sealedCap);
+		return nullptr;
+	}
+
+	Debug::log("Unsealed id: {} kind: {} size:{} item: {}",
+	           token->id,
+	           token->kind,
+	           token->maxSize,
+	           token->ConfigId);
+
+	if (token->id == 0)
+	{
+		// Assign an ID so we can track the callbacks added
+		// from this capability
+		static uint16_t nextId = 1;
+		token->id              = nextId++;
+	}
+
+	return token;
+}
+
+//
+// Find a Config by name.  If it doesn't already exist
+// create one.
+//
+NamedConfigItem *find_or_create_config(const char *name)
+{
+	for (auto &c : configData)
+	{
+		if (strcmp(c->name, name) == 0)
+		{
+			return c;
+		}
+	}
+
+	// Allocate a Config object
+	NamedConfigItem *c =
+	  static_cast<NamedConfigItem *>(malloc(sizeof(NamedConfigItem)));
+
+	// Save the name as a read only capability
+	auto nameBuffer = static_cast<char *>(malloc(strlen(name)));
+	strncpy(nameBuffer, name, strlen(name));
+	CHERI::Capability roName{nameBuffer};
+	roName.permissions() &=
+	  roName.permissions().without(CHERI::Permission::Store);
+	c->name         = roName;
+	c->item.version = 0;
+	c->item.data    = nullptr;
+
+	// Add it to the vector
+	configData.push_back(c);
+
+	return c;
+};
+
+//
+// Set a new value for the configuration item described by
+// the capability.
+//
+int __cheri_compartment("config_broker")
+  set_config(SObj sealedCap, void *data, size_t size)
+{
+	Debug::log("thread {} Set config called with {} {} {}",
+	           thread_id_get(),
+	           sealedCap,
+	           data,
+	           size);
+
+	ConfigToken *token = config_capability_unseal(sealedCap);
+	if (token == nullptr)
+	{
+		Debug::log("Invalid capability: {}", sealedCap);
+		return -1;
+	}
+
+	// Check we have a WriteToken
+	if (token->kind != WriteToken)
+	{
+		Debug::log(
+		  "Not a write capability for {}: {}", token->ConfigId, sealedCap);
+		return -1;
+	}
+
+	// Check the size and data are consistent with the token
+	// and each other.
+	if (size > token->maxSize)
+	{
+		Debug::log("invalid size {} for capability: {}", size, sealedCap);
+		return -1;
+	}
+
+	if (size > static_cast<size_t>(Capability{data}.bounds()))
```

Comment:
The header docs should be fairly complete. If I didn’t cover it in the book, can you raise an issue there?

The fast claims are for ephemeral reads of data that someone else owns. Each thread can fast-claim two objects, guaranteeing that they won’t be freed, until the next fast claim call or the next cross-compartment call.

Reply:
So its fast in the sense that it doesn't have to make a cross compartment call to the allocator (just a switcher library call), but its also different semantic compared to a heap_claim ?

If we're going to put a LockGuard in this call to avoid the race of multiple threads trying to  to set the same value does the risk of concurrent frees go away ?

Reply:
> So its fast in the sense that it doesn't have to make a cross compartment call to the allocator (just a switcher library call), but its also different semantic compared to a heap_claim ?

Indeed.  It's fast because it's not doing a cross-compartment call, not acquiring locks, not allocating memory to hold the claim metadata, and so on (`heap_claim` is quite an expensive operation).  The name is probably bad and we should give is something better such as `heap_claim_ephemeral`.

> If we're going to put a LockGuard in this call to avoid the race of multiple threads trying to to set the same value does the risk of concurrent frees go away ?

Not with the current API.  If a publisher has two threads, one can call the publish function in the broker, the other can concurrently free the object.  This is the reason we have the claim-fast, check-pointer, copy pattern.  It might be nice to add a `safe_memcpy` function to the compartment-helpers library that does this dance.

Reply:
Not sure I follow - the free in this case is inside the brokers set_config(), and that's going to be protected by the LockGuard, so if there were two publisher threads one of them will be waiting on the LockGuard won't it  ?